Port from `rs-wnfs`

expede opened the pull request.

expede committed:
Clean up from raw template

expede committed:
Fix tests

expede committed:
Add libipld for tests

expede committed:
de-win-ify

expede commented on flake.nix:
```diff
@@ -45,6 +45,23 @@
+      };
+
+      test-all =
+        pkgs.writeScriptBin "test:all" ''
+          #!${bin.bash}
+          echo "⚙️  Running cargo test"
+          ${bin.cargo} test
+
+          printf %"$COLUMNS"s | tr " " "-"
+          echo "⚙️  Running headless Wasm tests"
+          ${bin.wasm-pack} test --headless --chrome deterministic-bloom-wasm
+        '';
```
@zeeshanlakhani this just gets exported into the environment

<img width="698" alt="Screenshot 2022-12-19 at 20 44 16" src="https://user-images.githubusercontent.com/1052016/208585597-5b7c485f-8bd7-4b1a-b60c-d4862751f4c4.png">

zeeshanlakhani replied:
dope. want to add this to the rust-template flake for rust+wasm @expede? I'd typically do this in make (and we do this in ci), so I can see the benefit for nix peeps.

expede replied:
Sure, makes sense!

We have a much more elaborate version in the Haskell repo, but I want to clean that up and just publish a package at some stage.

zeeshanlakhani replied:
Ah nice. It’d have rusty things in it, but that’d
Be a nice add

expede committed:
fix: setup basics for tests

expede commented:
Licensing test is failing — will look at it in the morning, but it's otherwise RFR

zeeshanlakhani commented:
> Licensing test is failing — will look at it in the morning

just needs BSD-2-Clause and BSL-1.0 added to the list of licenses. I didn't want to make all the opines in the template.

expede commented:
Ah, clippy is unhappy about some stuff in the test utils. I'll look tomorrow; I have a flight so I have to get up relatively early

expede committed:
Fix licenses

zeeshanlakhani commented:
> > Licensing test is failing — will look at it in the morning
> 
> just needs BSD-2-Clause and BSL-1.0 added to the list of licenses. I didn't want to make all the opines in the template.

oh, derp, and CC0-1.0 :).

expede commented:
Oh also I wasn't really sure what to wrap for Wasm. I can serialize the filter, but that seems wasteful in the normal case 🤷‍♀️ I think I want to do some stuff with `WasmRefCell`s, but that seems like a "not this instant" thing. Thoughts?

expede committed:
Add CC0-1.0

expede committed:
Leaning the ways of the wasm

expede committed:
Leaning the ways of the wasm

expede committed:
Macro working

expede committed:
Switch to thiserror

expede committed:
LOL macro doctests

expede committed:
macro doctesting

expede committed:
WIP off to the holiday party I go

expede committed:
last minute typos

expede committed:
Test all macros

expede committed:
get wasm tests running again

appcypher commented on deterministic-bloom/src/lib.rs:
```diff
@@ -2,29 +2,361 @@
+    type Error = Error;
+
+    fn try_from(bytes: Vec<u8>) -> Result<Self, Self::Error> {
+        let bits = BitArray::<[u8; N]>::new(bytes.try_into().map_err(|vec: Vec<u8>| {
+            Error::VectorImportSizeMismatch {
+                expected: N,
+                actual: vec.len(),
+            }
+        })?);
```
Nice. Need to fix that in wnfs

appcypher commented on deterministic-bloom/tests/integration_test.rs:
```diff
@@ -1,4 +1,16 @@
 #[test]
-fn test_add() {
-    assert_eq!(deterministic_bloom::add(3, 2), 5);
+fn test_contains() {
+    let mut bloom = BloomFilter::<256, 30>::new();
+    let mut rvg = Rvg::new();
+
+    let new_val: Vec<u8> = vec![rvg.sample(&(..255u8))];
+    bloom.insert(&new_val);
+
+    for _ in 1..25 {
+        bloom.insert(&vec![rvg.sample(&(..255u8))]);
+    }
+
+    assert!(bloom.contains(&new_val));
 }
```
Would proptest macro work better here? I'm thinking @zeeshanlakhani wrote rvg for benches.
Just because the proptest macro is going to be more flexible.

```rs
// Not tested.

use test_strategy::proptest;
use proptest::collection::vec;

#[proptest(cases = 1000)]
fn test_contains(
    #[strategy(vec(vec(0..256u8, 0..100), 26))] 
    values: Vec<Vec<u8>>
) {
    let mut bloom = BloomFilter::<256, 30>::new();

    for v in vals {
        bloom.insert(&v);
    }
    
    assert!(bloom.contains(&vals[0]));
}
```

zeeshanlakhani replied:
yeah, purely for tests, we should abuse the proptest macro itself for something like this.

zeeshanlakhani commented on deterministic-bloom-wasm/src/lib.rs:
```diff
@@ -27,29 +23,167 @@ pub fn set_panic_hook() {
-macro_rules! console_log {
-    ($($t:tt)*) => ($crate::log(&format_args!($($t)*).to_string()))
-}
+gen_bloom!(SmallBloomFilter, 256, 13);
```
@expede do you think we'll want more in-between variations for wasm?

zeeshanlakhani commented on deterministic-bloom/src/lib.rs:
```diff
@@ -2,29 +2,361 @@
+}
+
+/// Errors for [BloomFilter] operations.
+#[derive(Error, Debug, Diagnostic)]
```
minor: I'd specify this as thiserror::Error (instead of the import above), just to be explicit in the derive case.

zeeshanlakhani commented on deterministic-bloom-wasm/src/lib.rs:
```diff
@@ -27,29 +23,167 @@ pub fn set_panic_hook() {
+
+        #[wasm_bindgen]
+        impl $name {
+            #[doc = concat!("Initialize a blank [", stringify!($name), "] (i.e. contains no elements).")]
```
wish there was a better way than all these sub-concat! and stringify! macros, but I guess that's the ticket, sadly.

zeeshanlakhani commented on README.md:
```diff
@@ -1,9 +1,5 @@
 <div align="center">
```
minor: replace expede as the org/user w/ wnfs-wg (throughout)

zeeshanlakhani commented on deterministic-bloom-wasm/src/lib.rs:
```diff
@@ -27,29 +23,167 @@ pub fn set_panic_hook() {
+/// Generate a monomorphic [BloomFilter] newtypes.
+macro_rules! gen_bloom {
+    ($name: ident, $n: expr, $k: expr) => {
+        #[doc = concat!("A monomorphic wrapper for [BloomFilter]`<", stringify!($n), ", ", stringify!($k), ">`.")]
```
For docs, I'd probably expand out what the $n and $k are more specifically, as it reads a bit weird when rendered w/ just the numbers within the angle brackets. 

<img width="997" alt="Screenshot 2023-01-09 at 10 57 48 AM" src="https://user-images.githubusercontent.com/202820/211351575-f637edcf-1268-4a96-ab8d-89003932aea6.png">
<img width="559" alt="Screenshot 2023-01-09 at 10 57 43 AM" src="https://user-images.githubusercontent.com/202820/211351581-c055fd78-faaa-40b1-bf2b-d263b3169114.png">

zeeshanlakhani commented on deterministic-bloom/tests/integration_test.rs:
```diff
@@ -1,4 +1,16 @@
 #[test]
-fn test_add() {
-    assert_eq!(deterministic_bloom::add(3, 2), 5);
+fn test_contains() {
```
does this cover all the tests (unit + integration) from the original port? I can also double check :).

matheus23 commented:
> Oh also I wasn't really sure what to wrap for Wasm. I can serialize the filter, but that seems wasteful in the normal case woman_shrugging I think I want to do some stuff with WasmRefCells, but that seems like a "not this instant" thing. Thoughts?

I'm thinking the most efficient way will be to expose it as `Rc<BloomFilter<...>>` and in some way make it so that it's available as a `Uint8Array` on the JS side. This would allow the `Uint8Array` to not be copied into the JS side, but instead the JS side would just have something that points into Wasm memory.

The `Rc` + something like wasm-bindgen's `--weak-refs` feature is how I imagine figuring out when repurposing the memory on the rust side is safe. I'm not 100% sure how that'd work.

matheus23 commented:
> do you think we'll want more in-between variations for wasm?

We'll probably want a runtime-dynamic version for Wasm, honestly.
The const-based version is good for doing benchmarks & WNFS, but it's not good for e.g. CAR mirror.

walkah commented:
@matheus23 any reason to not merge your PR #3 into this and merge this to main? I've created #4 as a follow up for dynamic sizing.

matheus23 commented:
@walkah no reason besides it's missing a review :man_shrugging:

matheus23 committed:
Port continuation (#3)

This PR addresses a bunch of Nits on the previous PR: #2 

Specifically:
- Instead of the integration test that tests a single sample, I added a
proptest.
- This :point_up_2: allowed me to remove the `Sampleable` stuff, since
that'd be duplicated between WNFS and this crate, which I don't think is
ideal. I'm not sure where to put the `Sampleable` trait best. Anyhow, we
don't really need it. Same with the `rvg` stuff.
- Changed any occurrences of `expede` in links/as the github owner to
`wnfs-wg`.
- Adjusted the docs in the -wasm crate to explain better what the const
params are.
- Some smaller nits such as how imports are made & some polishing.

matheus23 commented:
@walkah It's like I'm reviewing my own code haha
I'll just go ahead and merge
